File: src/myLeetCode/L275HIndexII.js
/**
 * @param {number[]} citations
 * @return {number}
 */
var hIndex = function(citations) {
    var len = citations.length;
    if (len<1) return 0;
    var b = 0, e = len-1, mid;
    var count = 0;
    while (b<=e){
    	mid = parseInt((b+e)/2);
    	count = len - mid;
    	if (citations[mid]>=count) e = mid-1;
    	else b = mid + 1;
    	
    }
    return len-(e+1);
};

var nums = [1,3,5,7,9];
console.log(hIndex(nums));
nums = [0,1];
console.log(hIndex(nums));
nums = [1,2,2,2];
console.log(hIndex(nums));

File: src/myLeetCode/L275HIndexII.java
package myLeetCode;

import java.util.Arrays;

import javax.xml.stream.events.EndDocument;

import com.sun.org.apache.xpath.internal.operations.And;

public class L275HIndexII {
	public int hIndex(int[] citations) {
        int len = citations.length;
        if (len<1) return 0;
        
        int b = 0, e = len-1, mid = (b+e)/2;
        int count = 0;
        int ans = 0;
        while (b<=e){
        	count = len - mid;
        	if (citations[mid] >= count)
        	{
        		if (citations[mid]==count) return count;
        		else e = mid-1;
        	}
        	else b = mid+1;
        	mid = (b+e)/2;
        }
        return len-(e+1);
        
        /*int i=0;
        for (i=0; i<len; i++){
        	if (citations[i]>=len-i) return len-i;
        }*/
        
        //return 0;
    }
	
	public void main(int[] nums){
		Arrays.sort(nums);
		System.out.println(hIndex(nums));
	}
}
